add an efficient way to add multiple items to a list at a time

radix opened the pull request:
Fixes #68 

First, a caveat: this is my very first Haskell contribution. My knowledge of idioms and conventions is very poor. It'd be *very* helpful to me if someone checked over this code and either told me how it should be changed or just made changes directly, so I can learn how to better write Haskell :)

This PR introduces two new functions:

- `addMultipleToList :: (Show b) => Widget (List a b) -> [(a, Widget b)] -> IO ()`
- `insertMultipleIntoList :: (Show b) => Widget (List a b) -> [(a, Widget b)] -> Int -> IO ()`

These maybe aren't as fast as they could be but I can now pretty much instantly add the ~100k elements instead of waiting for 30 seconds to do it with repeated calls to `addToList`.

Couple of extra notes:
- I'm really bad at arithmetic, apparently, but I think I've simplified the `newScrollTop` calculation and maintained its behavior, which is "scroll *just* enough to keep the selected item in view". (now that I say that, maybe there should be a separate exposed function that does exactly that?)
- the `addToList` documentation mentioned something about constructing widgets based on a widget constructor passed to newList, which I guess was an old API or one that never got into master? I've gotten rid of that clause.
- I also added a bunch of features to the List Demo, mostly to make sure I got all the edge cases right, and to test performance. If this is too much for a simple demo, I'll be happy to revert that part of the PR.

radix commented:
Hmm, it seems that these functions blow up when adding an empty list of items to an empty list widget. I'll fix.

radix commented:
ping @jtdaugherty -- just wanna make sure this wasn't forgotten :) It's all ready for review.

I suspect a similar optimization may be needed for tables (which I'll probably need to switch my application to using soon), but I can do that work separately from this PR.

jtdaugherty commented:
Hi. I'm listening and I've been reading updates from your tickets/PRs, but I haven't had time to think deeply about what's going on here so that's why I haven't commented yet.  However, I plan to do so as soon as I can. Thanks for your contributions!

Regarding tables, tables don't support scrolling, so while you can add rows to them, you won't be able to use tables instead of lists if you're depending on adding more things than can fit on the screen. However, you might want to take a look at the `scrolling` branch, since it provides a first pass at a generalized scrolling widget that makes it possible to scroll a table.  This is *not* finished and has *many* issues that I won't go into here, but it's something to try, anyway.

radix commented:
Thanks! I don't mean to rush you, just wanted to make sure you got a ping, since I wasn't sure if GitHub would notify you of new PRs without a @-mention.

jtdaugherty commented:
Yep, understood. Thanks! I've been meaning to comment to let you know that I was listening anyway. :)

jtdaugherty commented:
Also, would you mind taking a look at the section in the manual on the List widget to see if your API changes warrant any updates there, too?

jtdaugherty commented:
(And, to be clear, GitHub notifies me about everything that happens on the repo, regardless of mentions.)

jtdaugherty commented:
On Haskell style: thanks for being open to feedback. My main concerns are to ensure consistency with what whatever code is already in the library and to make sure contributions are comprehensible, decomposed, modular, etc.  (This might mean making *my* code better about those things.)  The PR you submitted looked fine to me, aside from the minor comments I left.

On the demos: more features in the demos are welcome, since they are supposed to show off one or two features of the library in a way that gives a good sense of what is possible.  They are also (hopefully) examples of relatively small, easy-to-understand programs to use to understand how to build a complete vty-ui program.

radix commented:
Thank you for the review @jtdaugherty! Sorry it took me so long to address your comments. The new version is pushed now.

radix commented:
Oh, I forgot about the user manual. I've never used LaTeX and I don't have the tools installed, but I can try to work it out.

radix commented:
Okay, I've also:

- renamed `widgets` to `pairs` for addMultipleToList (which I forgot in my first amendment)
- added a short paragraph about the new functions to the latex document.

I wasn't able to build the docs locally, so could you please doublecheck that the syntax (and prose) is okay?

I think that's everything!

jtdaugherty commented:
Thanks for the updates. I'll do a build of the manual to test, but it looks okay. If you end up installing LaTeX, you can build the manual by running `make -C doc`.

Thanks!

radix commented:
Yay! Thanks :sunglasses:
